OptimizelySDK.Tests/DecisionServiceTest.cs

```diff
@@ -441,5 +452,131 @@ public void TestGetVariationWithBucketingId()
             Assert.AreEqual(variationKeyControl, actualVariationKey, string.Format("Variation \"{0}\" does not match expected user profile variation \"{1}\".", actualVariationKey, variationKeyControl));
         }
+
+
+        #region GetVariationForFeature() method Tests
+
+        // Should return null and log a message when the feature flag's experiment ids array is empty
+        [Test]
+        public void TestGetVariationForFeatureExperimentGivenNullExperimentIds()
+        {
+            var featureFlag = ProjectConfig.GetFeatureFlagFromKey("empty_feature");
+
+            Assert.IsNull(DecisionService.GetVariationForFeatureExperiment(featureFlag, GenericUserId, new UserAttributes() { }));
+
+            LoggerMock.Verify(l => l.Log(LogLevel.INFO, $"The feature flag \"{featureFlag.Key}\" is not used in any experiments."));
+        }
+
+        // Should return null and log when the experiment is not in the datafile 
+        [Test]
+        public void TestGetVariationForFeatureExperimentGivenExperimentNotInDataFile()
+        {
+            var featureFlagWithInvalidExperiment = new FeatureFlag { Id = "29039203", Key = "boolean_feature", RolloutId = "abc", ExperimentIds = new List<string> { "29039203" } };
+
+            Assert.IsNull(DecisionService.GetVariationForFeatureExperiment(featureFlagWithInvalidExperiment, GenericUserId, new UserAttributes() { }));
+
+            LoggerMock.Verify(l => l.Log(LogLevel.ERROR, "Experiment ID \"29039203\" is not in datafile."));
+        }
+
+        // Should return nulll and log when the user is not bucketed into the feature flag's experiments
+        [Test]
+        public void TestGetVariationForFeatureExperimentGivenNonMutexGroupAndUserNotBucketed()
+        {
+            var multiVariateExp = ProjectConfig.GetExperimentFromKey("test_experiment_multivariate");
+
+            DecisionServiceMock.Setup(ds => ds.GetVariation(multiVariateExp, "user1", null)).Returns<Variation>(null);
+            var featureFlag = ProjectConfig.GetFeatureFlagFromKey("multi_variate_feature");
+
+            Assert.IsNull(DecisionServiceMock.Object.GetVariationForFeatureExperiment(featureFlag, "user1", new UserAttributes()));
+
+            LoggerMock.Verify(l => l.Log(LogLevel.INFO, "The user \"user1\" is not bucketed into any of the experiments on the feature \"multi_variate_feature\"."));
+        }
+
+        // Should return the variation when the user is bucketed into a variation for the experiment on the feature flag
+        [Test]
+        public void TestGetVariationForFeatureExperimentGivenNonMutexGroupAndUserIsBucketed()
+        {
+            var expectedVariation = ProjectConfig.GetVariationFromId("test_experiment_multivariate", "122231");
+
+            DecisionServiceMock.Setup(ds => ds.GetVariation(ProjectConfig.GetExperimentFromKey("test_experiment_multivariate"), "user1", new UserAttributes())).Returns(expectedVariation);
```

Comment:
new UserAttributes will allocate new memory and it mocking method won't be called. Please place in the variable and use that reference at line 501 and 504.

Reply:
Done